src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/CastHelpers.cs

```diff
@@ -21,6 +21,7 @@ namespace System.Runtime.CompilerServices
     internal static unsafe class CastHelpers
     {
         private static int[]? s_table;
+        private static readonly MethodTable* pObjMt = (MethodTable*)RuntimeTypeHandle.GetValueInternal(typeof(object).TypeHandle);
```

Comment:
I do not think you need to cache it in a static if the intrinsic expansion works the way it is supposed to work.

Also, note that adding a before field init static constructor will make access to `s_table` slower without tiered JIT.

Reply:
Going with:

```cs
            notExactMatch:
                if (arr.GetType() == typeof(object[]))
                    goto doWrite;
```

tier JIT-ed:
```
            notExactMatch:
                if (array.GetType() == typeof(object[]))
00007FFAD1611A7B  mov         rdx,7FFAD15DADD8h  
00007FFAD1611A85  cmp         qword ptr [rcx],rdx  
00007FFAD1611A88  je          00007FFAD1611A62  
                    goto doWrite;

```
w tiering disabled
```
            notExactMatch:
                if (array.GetType() == typeof(object[]))
00007FFB294A8EAE  mov         rcx,qword ptr [rcx]  
00007FFB294A8EB1  cmp         rcx,qword ptr [7FFB28F7C6A8h]  
00007FFB294A8EB8  je          00007FFB294A8E92  
                    goto doWrite;


```